drivers/counter/counter_mcux_ctimer.c

```diff
@@ -0,0 +1,241 @@
+
+	if (data->channels[chan_id].alarm_callback != NULL) {
+		return -EBUSY;
+	}
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
+		ticks += current;
+	}
+
+	data->channels[chan_id].alarm_callback = alarm_cfg->callback;
+	data->channels[chan_id].alarm_user_data = alarm_cfg->user_data;
+
+	ctimer_match_config_t match_config = { .matchValue = ticks,
+					       .enableCounterReset = false,
+					       .enableCounterStop = false,
+					       .outControl = kCTIMER_Output_NoAction,
+					       .outPinInitState = false,
+					       .enableInterrupt = true };
+
+	CTIMER_SetupMatch(config->base, chan_id, &match_config);
+
+	return 0;
+}
+
+static int mcux_lpc_ctimer_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+	struct mcux_lpc_ctimer_data *data = dev->data;
+
+	CTIMER_DisableInterrupts(config->base, (1 << chan_id));
+
+	data->channels[chan_id].alarm_callback = NULL;
+	data->channels[chan_id].alarm_user_data = NULL;
+
+	return 0;
+}
+
+static int mcux_lpc_ctimer_set_top_value(const struct device *dev,
+					 const struct counter_top_cfg *cfg)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(cfg);
+	return -ENOTSUP;
+}
+
+static uint32_t mcux_lpc_ctimer_get_pending_int(const struct device *dev)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+
+	return (CTIMER_GetStatusFlags(config->base) & 0xF) != 0;
+}
+
+static uint32_t mcux_lpc_ctimer_get_top_value(const struct device *dev)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+
+	return config->info.max_top_value;
+}
+
+static void mcux_lpc_ctimer_isr(const struct device *dev)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+	struct mcux_lpc_ctimer_data *data = dev->data;
+
+	uint32_t interrupt_stat = CTIMER_GetStatusFlags(config->base);
+
+	CTIMER_ClearStatusFlags(config->base, interrupt_stat);
+
+	uint32_t ticks = mcux_lpc_ctimer_read(config->base);
+
+	for (uint8_t chan = 0; chan < NUM_CHANNELS; chan++) {
+		uint8_t channel_mask = 0x01 << chan;
+
+		if (((interrupt_stat & channel_mask) != 0) &&
+		    (data->channels[chan].alarm_callback != NULL)) {
+			counter_alarm_callback_t alarm_callback =
+				data->channels[chan].alarm_callback;
+			void *alarm_user_data = data->channels[chan].alarm_user_data;
+
+			data->channels[chan].alarm_callback = NULL;
+			data->channels[chan].alarm_user_data = NULL;
+			alarm_callback(dev, chan, ticks, alarm_user_data);
+		}
+	}
+}
+
+static int mcux_lpc_ctimer_init(const struct device *dev)
+{
+	/*
+	 * The frequency of the timer is not known at compile time so we need to
+	 * modify the timer's config in the init function at runtime when the
+	 * frequency is known.
+	 */
+	struct mcux_lpc_ctimer_config *config = (struct mcux_lpc_ctimer_config *)dev->config;
+
+	ctimer_config_t ctimer_config;
+
+	CLOCK_AttachClk(config->clock_source);
+
+	uint32_t clk_freq = CLOCK_GetCTimerClkFreq(config->timer);
```

Comment:
Please move AttachClk to soc.c, and use the zephyr clock control driver (`drivers/clock_control/clock_control_mcux_syscon.c`) to get the frequency.

Reply:
Change to use the clock control API.